truffle/src/main/java/org/jruby/truffle/runtime/object/ObjectGraph.java

```diff
@@ -14,190 +14,155 @@
-                } else {
-                    return false;
-                }
-            }
-
-        });
-
-        return objects;
-    }
+            public void run(DynamicObject thread, Node currentNode) {
+                synchronized (visited) {
+                    final Deque<DynamicObject> stack = new ArrayDeque<>();
 
-    private void visitRoots(final ObjectGraphVisitor visitor) {
-        final Thread mainThread = Thread.currentThread();
+                    stack.add(thread);
 
-        context.getSafepointManager().pauseAllThreadsAndExecute(null, false, new SafepointAction() {
+                    if (Thread.currentThread() == stoppingThread) {
+                        stack.addAll(ObjectGraph.getAdjacentObjects(context.getCoreLibrary().getGlobalVariablesObject()));
+                        stack.addAll(context.getAtExitManager().getHandlers());
+                        stack.addAll(context.getObjectSpaceManager().getFinalizerHandlers());
+                    }
 
-            boolean keepVisiting = true;
+                    final FrameInstance currentFrame = Truffle.getRuntime().getCurrentFrame();
 
-            @Override
-            public void run(DynamicObject thread, Node currentNode) {
-                synchronized (this) {
-                    if (!keepVisiting) {
-                        return;
+                    if (currentFrame != null) {
+                        stack.addAll(getObjectsInFrame(currentFrame.getFrame(FrameInstance.FrameAccess.READ_ONLY, false)));
                     }
 
-                    try {
-                        // We only visit the global variables and other global state from the root thread
+                    Truffle.getRuntime().iterateFrames(new FrameInstanceVisitor<Object>() {
 
-                        if (Thread.currentThread() == mainThread) {
-                            visitObject(context.getCoreLibrary().getGlobalVariablesObject(), visitor);
+                        @Override
+                        public Object visitFrame(FrameInstance frameInstance) {
+                            stack.addAll(getObjectsInFrame(frameInstance
+                                    .getFrame(FrameInstance.FrameAccess.READ_ONLY, false)));
 
-                            for (DynamicObject handler : context.getAtExitManager().getHandlers()) {
-                                visitObject(handler, visitor);
-                            }
-
-                            for (DynamicObject handler : context.getObjectSpaceManager().getFinalizerHandlers()) {
-                                visitObject(handler, visitor);
-                            }
+                            return null;
                         }
 
-                        // All threads visit their thread object
-
-                        visitObject(thread, visitor);
+                    });
 
-                        // All threads visit their call stack
+                    while (!stack.isEmpty()) {
```

Comment:
I use an explicit stack object rather than the Java callstack as object graphs could be arbitrarily broad and we don't want a stack overflow.